File: learnOpencv/Final1/c++/main.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include <opencv2/imgcodecs.hpp> 
#include <opencv2/opencv.hpp>
#include<iostream>

using namespace std;
using namespace cv;

int main()
{
    Mat image = imread("../c++/52.jpg");
    Mat image2 = image.clone();
    int height = image.rows;
    int width = image.cols;
    Mat cir = Mat::zeros(height, width, CV_8UC3);
    imshow("initial", image);
    cvtColor(image, image, COLOR_BGR2GRAY);
    GaussianBlur(image, image, Size(9,9), 2, 2);
    vector<Vec3f> circles;
    HoughCircles(image, circles, HOUGH_GRADIENT, 1, 5, 100, 40);
    cout << "Number of circle is : " << circles.size() << endl;
    cvtColor(image, image, COLOR_GRAY2BGR);
    for(size_t i = 0; i < circles.size(); i++)
    {
        Vec3f cc = circles[i];
        Point center(cvRound(cc[0]), cvRound(cc[1]));
        int radius = cvRound(cc[2]);
        Scalar color1(0, 255, 255);
        circle(image2, center, radius, color1, -1);
    }    
    imshow("change", image2);
    Mat black = Mat::zeros(image.size(), CV_8UC3);
    for (size_t i = 0; i < circles.size(); i++) {
        Vec3f cc = circles[i];
        Point center(cvRound(cc[0]), cvRound(cc[1]));
        int radius = cvRound(cc[2]);
        Scalar color1(255, 255, 255);
        circle(black, center, radius, color1, -1);
    }
    Mat ROI;
    bitwise_and(image, black, ROI);
    imshow("ROI", ROI);
    waitKey(0);
    return 0;
}

File: learnOpencv/Final1/python/main.py
import cv2 as cv
import numpy as np

image = cv.imread("../python/52.jpg")
image2 = cv.imread("./52.jpg")
height = image.shape[0]
width = image.shape[1]
cir = np.zeros((height, width, 3), np.uint8)
cv.imshow("initial", image)

gray_image = cv.cvtColor(image, cv.COLOR_BGR2GRAY)
gray_image = cv.GaussianBlur(gray_image, (9, 9), 2, 2)
circles = cv.HoughCircles(gray_image, cv.HOUGH_GRADIENT, 1, 5, param1=100, param2=40)
print("Number of circles: ", circles.shape[1])
circles = np.uint16(circles)
for i in circles[0, :]:
    center = (i[0], i[1])
    radius = i[2]
    color = (0, 255, 255)
    cv.circle(image, center, radius, color, -1)
cv.imshow("circles", image)
for i in circles[0, :]:
    center = (i[0], i[1])
    radius = i[2]
    color = (255, 255, 255)
    cv.circle(cir, center, radius, color, -1)
ROI = cv.bitwise_and(image2, cir)
cv.imshow("ROI", ROI)
cv.waitKey(0)